cpp/src/arrow/compute/row/doc/row_table.md

```diff
@@ -0,0 +1,84 @@
+
+For example, a row table with the schema `(int32, boolean)` and rows `[[7, false], [8, true], [9, false], ...]` is stored in the fixed-length buffer as follows:
+| Row 0                | Row 1                | Row 2                | ... |
+|----------------------|----------------------|----------------------|-----|
+| 7 0 0 0, 0 (padding) | 8 0 0 0, 1 (padding) | 9 0 0 0, 0 (padding) | ... |
+
+### Offsets for Varying-length Row Data
+
+In a varying-length row table, the fixed-length buffer contains offsets to the varying-length row data, which is stored separately in the optional varying-length buffer. The offsets are of type `RowTableMetadata::offset_type` (fixed as `int64_t`) and indicate the starting position of the row data for each row.
+
+### Varying-length Row Data
+
+In a varying-length row table, the varying-length buffer contains the actual row data, stored contiguously. The offsets in the fixed-length buffer point to the starting position of each row's data.
+
+#### Row Encoding
+
+A varying-length row is encoded as follows:
+- Fixed-length columns are stored first.
+- A sequence of offsets to each varying-length column follows. Each offset is 32-bit and indicates the **end** position within the row data of the corresponding varying-length column.
+- Varying-length columns are stored last.
+
+For example, a row table with the schema `(int32, string, string, int32)` and rows `[[7, 'Alice', 'x', 0], [8, 'Bob', 'y', 1], [9, 'Charlotte', 'z', 2], ...]` is stored as follows (assuming 8-byte alignment for varying-length columns):
+
+Fixed-length buffer (row offsets):
+| Row 0           | Row 1            | Row 2            | Row 3             | ... |
+|-----------------|------------------|------------------|-------------------|-----|
+| 0 0 0 0 0 0 0 0 | 32 0 0 0 0 0 0 0 | 64 0 0 0 0 0 0 0 | 104 0 0 0 0 0 0 0 | ... |
+
+Varying-length buffer (row data):
+| Row | Fixed-length Cols | Varying-length Offsets | Varying-length Cols        |
+|-----|-------------------|------------------------|----------------------------|
+| 0   | 7 0 0 0, 0 0 0 0  | 21 0 0 0, 25 0 0 0     | `Alice~~~x~~~~~~~`         |
+| 1   | 8 0 0 0, 1 0 0 0  | 19 0 0 0, 25 0 0 0     | `Bob~~~~~y~~~~~~~`         |
+| 2   | 9 0 0 0, 2 0 0 0  | 25 0 0 0, 33 0 0 0     | `Charlotte~~~~~~~z~~~~~~~` |
+| 3   | ...               | ...                    | ...                        |
+
+### Alignment
+
+Each row in the row table is aligned to `RowTableMetadata::row_alignment` bytes. Fixed-length columns with non-power-of-2 lengths are also aligned to `RowTableMetadata::row_alignment` bytes. Varying-length columns are aligned to `RowTableMetadata::string_alignment` bytes.
```

Comment:
Is `row_alignment` configurable at runtime or is it a compile-time constant?

Reply:
It can be configured by specifying this field before constructing a row table:
https://github.com/apache/arrow/blob/520ae44272d491bbb52eb3c9b84864ed7088f11a/cpp/src/arrow/compute/row/row_internal.h#L65-L66

Reply:
Two follow-ons to this section:

- Can we give the developer any guidance on whether what value they should pick?
- Does row_alignment always need to be the same between the Row Tables (like in AppendSelectionFrom)?